### DISCUSSION TITLE
how to back pressure correctly?

### DISCUSSION THREAD

**Discussion by dominictarr** [libuv/libuv#3434] [Q&A]
I'm trying to understand how tcp back pressure should be implemented with libuv.

I've been reading the libuv code, so I understand things like that there is an internal queue of write requests.
Also, I am experienced with node.js streams, and I have attempted to read the node code, but it is too complicated to make sense of. (Curiously, the c++ side of node.js is actually larger than libuv!)

Anyway, I understand that libuv calls the os write method, and if it can take it, it does, and returns immediately - so one way would be to use `uv_try_write` and apply back pressure if it fails. This would avoid using libuv's write queue.

(interestingly, node.js also avoids using the libuv write queue, it only allows one writev call in process at a time - I think this is to avoid too many calls between js and c++ which are understood to be expensive)

I can see that using `uv_send_buffer_size` and `uv_recv_buffer_size` it's possible to query the operating system's buffer sizes, but there is no way to set the sizes, but I understand that on \*nix at least, you can set the so buffer size using `setsockopt`. Did libuv leave this out for cross platform reasons?

So I guess the right method would be to choose my own buffer size, and then check `uv_get_write_queue_size` to see how much is left in it and ignore the system buffer?

**Comment by bnoordhuis** [comment]
The easy way to implement back-pressure is to have a single `uv_write_t` request in flight, and only start a new write when the previous one completed.

Apropos `uv_try_write()`: you have to take care to avoid starvation, i.e., don't blindly keep writing as long as it succeeds because fast peers are going to starve the event loop.

>  can see that using uv_send_buffer_size and uv_recv_buffer_size it's possible to query the operating system's buffer sizes, but there is no way to set the sizes,

Those functions let you get and set the buffer sizes, it depends on the `value` in/out parameter.

**Comment by dominictarr** [comment]
thanks for your response - but can you clarify what starve means? "don't blindly keep writing as long as it succeeds because fast peers are going to _starve_ the event loop" does it mean that requests on that that handle will get processed in preference to other requests?

> it depends on the value in/out parameter.

oh I see now, the value is passed as a pointer to an int.

**Reply by bnoordhuis** [reply]
Starving is when you block the event loop and stop other tasks from making progress.

It's only an issue with `uv_try_write()` because then you're opting into synchronous-when-possible behavior. `uv_write()` is always async.
